### PR TITLE
Fix patchcore image-level score computation

### PR DESCRIPTION
**Pull request by djdameln**
# Description

- This PR fixes the incorrect image-level anomaly score computation of PatchCore. Also fixes the issue that test batch sizes bigger than 1 were not supported.

- Fixes https://github.com/openvinotoolkit/anomalib/issues/286, https://github.com/openvinotoolkit/anomalib/issues/268, #533 

## Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/patchcore/torch_model.py [hunk 1]
```diff
@@ -134,7 +142,7 @@ def subsample_embedding(self, embedding: torch.Tensor, sampling_ratio: float) ->
         coreset = sampler.sample_coreset()
         self.memory_bank = coreset
 
-    def nearest_neighbors(self, embedding: Tensor, n_neighbors: int = 9) -> Tensor:
+    def nearest_neighbors(self, embedding: Tensor, n_neighbors: int = 9) -> Tuple[Tensor, Tensor]:
```
Comment (samet-akcay) [inline anomalib/models/patchcore/torch_model.py:145]: maybe we could remove the default value for `n_neighbors`?

File: anomalib/models/patchcore/torch_model.py [hunk 2]
```diff
@@ -143,7 +151,35 @@ def nearest_neighbors(self, embedding: Tensor, n_neighbors: int = 9) -> Tensor:
...
             Tensor: Patch scores.
+            Tensor: Locations of the nearest neighbor(s).
         """
         distances = torch.cdist(embedding, self.memory_bank, p=2.0)  # euclidean norm
-        patch_scores, _ = distances.topk(k=n_neighbors, largest=False, dim=1)
-        return patch_scores
+        patch_scores, locations = distances.topk(k=n_neighbors, largest=False, dim=1)
+        return patch_scores, locations
+
+    def compute_anomaly_score(self, patch_scores: torch.Tensor, locations: Tensor, embedding: Tensor) -> torch.Tensor:
```
Comment (samet-akcay) [inline anomalib/models/patchcore/torch_model.py:160]: for the sake of consistency, `torch.Tensor` could be annotated as `Tensor` maybe?

**Review by samet-akcay — APPROVED** [review]
Thanks a lot for the fix! It's been waiting for too long

**Review by samet-akcay — APPROVED** [review]
Thanks!

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks. Much cleaner as well!

**Comment by cyj95** [comment]
图片
pixel level score is still lower than official version.
for example.
“wood” category，your version：92.8，official version：95.1

**Comment by watertianyi** [comment]
@samet-akcay @cyj95 
patchcore uses batchsize for inference, and the time consumption has not been reduced. Is this because patchcore can only infer one image at a time?

**Pull request merged**